kiwmi/desktop/output.c

```diff
@@ -376,8 +390,129 @@ new_output_notify(struct wl_listener *listener, void *data)
     wl_signal_emit(&desktop->events.new_output, output);
 }
 
+void
+output_layout_change_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_desktop *desktop =
+        wl_container_of(listener, desktop, output_layout_change);
+    struct wlr_output_configuration_v1 *configuration =
+        wlr_output_configuration_v1_create();
+
+    struct kiwmi_output *output;
+    wl_list_for_each (output, &desktop->outputs, link) {
+        struct wlr_output_configuration_head_v1 *head =
+            wlr_output_configuration_head_v1_create(
+                configuration, output->wlr_output);
+
+        struct wlr_box *area = wlr_output_layout_get_box(
+            desktop->output_layout, output->wlr_output);
+
+        head->state.enabled = output->wlr_output->enabled;
+        head->state.mode    = output->wlr_output->current_mode;
+        if (area != NULL) {
+            head->state.x = area->x;
+            head->state.y = area->y;
+        }
+    }
+
+    wlr_output_manager_v1_set_configuration(
+        desktop->output_manager, configuration);
+}
+
+static void
+output_manager_configure(
+    struct kiwmi_server *server,
+    struct wlr_output_configuration_v1 *configuration,
+    bool test)
+{
+    bool okay = true;
+    struct wlr_output_configuration_head_v1 *head;
+    wl_list_for_each (head, &configuration->heads, link) {
+        struct wlr_output *wlr_output = head->state.output;
+        struct kiwmi_output *output   = wlr_output->data;
+
+        output->enable_changed = head->state.enabled != wlr_output->enabled;
+
+        wlr_output_enable(wlr_output, head->state.enabled);
+        if (head->state.enabled) {
+            if (head->state.mode) {
+                wlr_output_set_mode(wlr_output, head->state.mode);
+            } else {
+                wlr_output_set_custom_mode(
+                    wlr_output,
+                    head->state.custom_mode.width,
+                    head->state.custom_mode.height,
+                    head->state.custom_mode.refresh);
+            }
+
+            wlr_output_layout_move(
+                server->desktop.output_layout,
+                wlr_output,
+                head->state.x,
+                head->state.y);
+            wlr_output_set_transform(wlr_output, head->state.transform);
+            wlr_output_set_scale(wlr_output, head->state.scale);
+        }
+
+        if (!wlr_output_test(wlr_output)) {
+            okay = false;
+            break;
+        }
+    }
+
+    wl_list_for_each (head, &configuration->heads, link) {
+        if (okay && !test) {
+            struct kiwmi_output *output = head->state.output->data;
+            if (head->state.enabled) {
+                if (output->enable_changed) {
```

Comment:
Lol, i’m dumb. Of course would i focus on writing those mostly worthless nitpicks and forget that if `enable_changed`, the output should be added to / removed from the `output_layout` to match the behaviour of other compositors. (And, because we don’t want to send intermediate state, the `output_layout_change` listener should be disabled / no-op while the configuration is in progress. Which also somewhat replies to https://github.com/buffet/kiwmi/pull/62#discussion_r907448108.)

Reply:
(This one’s also resolved with the 2022-07-11 force-push; the parenthesised part lives on in another comment below.)